{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "313611ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gcsfs\n",
    "print(gcsfs.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bda3c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, io, sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a1341e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import base64\n",
    "import vertexai\n",
    "from vertexai.generative_models import GenerativeModel, Part, SafetySetting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7656c43",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML, Markdown\n",
    "\n",
    "def set_css_in_cell_output(unused):\n",
    "  display(HTML(\"\"\"\"\"\"))\n",
    "\n",
    "get_ipython().events.register('pre_run_cell', set_css_in_cell_output)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a02dec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Synapse client for data access\n",
    "import synapseclient\n",
    "from synapseclient import Synapse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34f88dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "!synapse login -p eyJ0eXAiOiJKV1QiLCJraWQiOiJXN05OOldMSlQ6SjVSSzpMN1RMOlQ3TDc6M1ZYNjpKRU9VOjY0NFI6VTNJWDo1S1oyOjdaQ0s6RlBUSCIsImFsZyI6IlJTMjU2In0.eyJhY2Nlc3MiOnsic2NvcGUiOlsidmlldyIsImRvd25sb2FkIiwibW9kaWZ5Il0sIm9pZGNfY2xhaW1zIjp7fX0sInRva2VuX3R5cGUiOiJQRVJTT05BTF9BQ0NFU1NfVE9LRU4iLCJpc3MiOiJodHRwczovL3JlcG8tcHJvZC5wcm9kLnNhZ2ViYXNlLm9yZy9hdXRoL3YxIiwiYXVkIjoiMCIsIm5iZiI6MTc2MzUxNjA5NCwiaWF0IjoxNzYzNTE2MDk0LCJqdGkiOiIyODc3MSIsInN1YiI6IjM1NjQ3ODEifQ.oC_oWQdMz6Qvje7CL_YVvjooe9fqsaGE56CiBQiSQwJ1fwUShjWhroTVMlaPEloTeyhd-h_tJy27YGuFOlbwOG0f7d27rkI3J42aTF1fEcKc_bPcwFmTkUlUomiYVUVd1XW1ECTFp6PCQUigWiHcEgINMI_R877b--Adj2VOlbgKt0kuIbSgLV0oadDhaUhtZEdKzCTN0m7Z1ZRFjOaDocaCVwgJWjpJ6lVQCPfsjipbE5ex8V1SY2Xxu76kGeYL3neoyRB7adZiV_fsn1Tyj0RK3ttWhwyoPIV37R02x5OdJ7u7D5JQW_c82oseUyA0ewb0YAywPxdzbFK9LtEsXw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198de97b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Google BigQuery and data handling\n",
    "import pandas as pd\n",
    "from google.cloud import bigquery\n",
    "import pandas_gbq  # To read BigQuery data into pandas DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e918ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the path to the uploaded file\n",
    "# metadata_df = pd.read_csv('/content/HTAN_ST_metadata_ad.tsv', sep='\\t')  # Use sep='\\t' for TSV files\n",
    "df = pd.read_csv('ST_data.tsv', sep='\\t') \n",
    "\n",
    "metadata_df = df.iloc[:,[0,2]]\n",
    "# Display the dataframe\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3756dab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_ID = \"ceremonial-port-478702-v2\"\n",
    "LOCATION = \"europe-north1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "842409da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vertexai\n",
    "vertexai.init(project=PROJECT_ID, location=LOCATION)\n",
    "from vertexai.preview import reasoning_engines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea31b741",
   "metadata": {},
   "outputs": [],
   "source": [
    "generation_config = {\n",
    "    # \"transport\":\"rest\",\n",
    "    \"max_output_tokens\": 8192,\n",
    "    \"temperature\": 0.5,\n",
    "    \"top_p\": 0.95,\n",
    "}\n",
    "\n",
    "safety_settings = [\n",
    "    SafetySetting(\n",
    "        category=SafetySetting.HarmCategory.HARM_CATEGORY_HATE_SPEECH,\n",
    "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
    "    ),\n",
    "    SafetySetting(\n",
    "        category=SafetySetting.HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,\n",
    "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
    "    ),\n",
    "    SafetySetting(\n",
    "        category=SafetySetting.HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,\n",
    "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
    "    ),\n",
    "    SafetySetting(\n",
    "        category=SafetySetting.HarmCategory.HARM_CATEGORY_HARASSMENT,\n",
    "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
    "    ),\n",
    "]\n",
    "\n",
    "def generate(question = \"\"\"What can you do?\"\"\"):\n",
    "\n",
    "    ## General prompt\n",
    "    prompt = f\"\"\"\n",
    "    You are a bioinformatics expert coder using the Human Tumor Network Atlas (HTAN)\n",
    "    datasets via Google Cloud and Jupyter. You can write Python code to answer\n",
    "    questions by writing BigQuery queries, pandas, scanpy, and squidpy code.\n",
    "    When a user asks a question, identify whether you need to write any code to\n",
    "    answer.\n",
    "    Very important: !! For any general questions and conversations,\n",
    "    you can answer with a conversation. For any questions that needs the code,\n",
    "    just write and return the Python code enclosed in ```python``` tags so that\n",
    "    the user can regex extract the code easily.\n",
    "    You do not have to access any data. You are an expert coder, just write the\n",
    "    code. When you're asked to load a dataset, use BigQuery to load it.\n",
    "\n",
    "    Include all the necessary import packages to run the code. If using any\n",
    "    Google Cloud services, you can use:\n",
    "    project_id = \"isb-cgc-external-004\"\n",
    "\n",
    "    Here are some examples:\n",
    "\n",
    "    Question: Can you load the cells from the HTAN biospecimen HTA7_1_3?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "    WITH cells AS (\n",
    "      SELECT  CellID, X_centroid, Y_centroid,\n",
    "      FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
    "      WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
    "    )\n",
    "    SELECT CellID, X_centroid,  Y_centroid\n",
    "    FROM cells\n",
    "    '''\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "\n",
    "    #### Synapse data download : Example 1\n",
    "\n",
    "    prompt += f\"\"\"\n",
    "    Question: Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current\n",
    "    where the File_Format is hdf5?\n",
    "\n",
    "    Answer:\n",
    "    '''\n",
    "    import pandas_gbq\n",
    "    from google.cloud import bigquery\n",
    "\n",
    "    project_id = \"isb-cgc-external-004\"\n",
    "\n",
    "    query = '''\n",
    "    SELECT entityId\n",
    "    FROM `isb-cgc-bq.HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current`\n",
    "    WHERE File_Format = 'hdf5'\n",
    "    '''\n",
    "\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ##  Synapse data download example 2\n",
    "\n",
    "    prompt += f\"\"\"\n",
    "    Question: Can you download the synapse data 'syn51133602' to /content/datasets/?\n",
    "    Answer:\n",
    "    '''\n",
    "    import synapseclient\n",
    "    syn = synapseclient.login()\n",
    "    entity = syn.get('syn51133602', downloadLocation='/content/datasets')\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ### Spatial tx example 1\n",
    "\n",
    "    prompt += f\"\"\"\n",
    "    Question: Can you categorize cells within a defined spatial region as either 'Tumor' or 'Other' based on threshold expression levels of specific markers (SOX10_cellRingMask, S100B_cellRingMask, and CD63_cellRingMask)?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "      WITH cells AS (\n",
    "      SELECT CellID, X_centroid, Y_centroid,\n",
    "      IF (SOX10_cellRingMask > 3704.5 AND (S100B_cellRingMask > 7589.48 OR CD63_cellRingMask > 570.68),\n",
    "      'Tumor', 'Other') AS celltype\n",
    "      FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
    "      WHERE HTAN_Biospecimen_ID = 'HTA7_1_3')\n",
    "\n",
    "      SELECT CellID, X_centroid, Y_centroid, celltype\n",
    "      FROM cells\n",
    "      WHERE X_centroid > 23076.9 AND X_centroid < 30384.6\n",
    "      AND Y_centroid > 9615.3 AND Y_centroid < 15000\n",
    "      '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ### Spatial tx example 2\n",
    "    prompt += f\"\"\"\n",
    "    Question: Can you classify cells as 'Tumor' or 'Other', convert their pixel coordinates to geospatial points, and calculate distances between cell pairs that are within a 20-micrometer proximity threshold\n",
    "    Answer:\n",
    "    query = '''\n",
    "      WITH geodat AS (\n",
    "      SELECT CellID, X_centroid, Y_centroid,\n",
    "      IF (SOX10_cellRingMask > 3704.5 AND (S100B_cellRingMask > 7589.48 OR CD63_cellRingMask > 570.68),\n",
    "      'Tumor', 'Other') AS celltype,\n",
    "      ST_GeogPoint(X_centroid / 368570, Y_centroid / 368570) AS p\n",
    "      FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
    "      WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
    "      )\n",
    "    SELECT t1.CellID, t1.X_centroid, t1.Y_centroid, t1.p, t1.celltype,\n",
    "    t2.CellID AS CellID_1, t2.X_centroid AS X_centroid_1, t2.Y_centroid AS Y_centroid_1, t2.p AS p_1, t2.celltype AS celltype_1,\n",
    "    ST_Distance(t1.p, t2.p) AS Distance\n",
    "    FROM geodat AS t1\n",
    "    JOIN geodat AS t2\n",
    "    ON ST_DWithin(t1.p, t2.p, 9.29324770787722)\n",
    "    '''\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ### Spatial tx example 3\n",
    "    prompt += f\"\"\"\n",
    "    Question: For each tumor cell within the specified spatial region, calculate the number of neighboring tumor cells\n",
    "    Answer:\n",
    "    query = '''\n",
    "      WITH cellp AS (\n",
    "      SELECT CellID, celltype, CellID_1, celltype_1\n",
    "      FROM `isb-cgc-bq.temp15432.Melanoma_CyCIF_HTA7_1_3_points_within_20um`\n",
    "      WHERE X_centroid > 23076.9 AND X_centroid < 30384.6\n",
    "      AND Y_centroid > 9615.3 AND Y_centroid < 15000)\n",
    "\n",
    "      SELECT CellID, COUNTIF(celltype_1 = 'Tumor') - 1 AS N_Tumor_Cells\n",
    "      FROM cellp\n",
    "      WHERE celltype = 'Tumor'\n",
    "      GROUP BY CellID\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    ## Metadata\n",
    "    # Create a prompt with the file content included\n",
    "    prompt += f\"\"\"\n",
    "    The HTAN metadata contains a subset of these type of attributes. Also included\n",
    "    is a description of these attributes. Internalize this information and use it\n",
    "    to answer any queries related to metadata.\n",
    "\n",
    "    Metadata Description:\n",
    "    {metadata_df}\n",
    "\n",
    "    Question: Which attributes describe the spatial location of cells?\n",
    "    Answer:\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "\n",
    "    ## ScRNAseq example 1\n",
    "    prompt += f\"\"\"\n",
    "    Question: What are the counts of unique cells, sex groupings, samples,\n",
    "    cell types, and therapies by development stage in the MSK scRNAseq dataset?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "    SELECT\n",
    "      development_stage,\n",
    "      count(distinct(iObs)) AS Number_Cells,\n",
    "      count(distinct(sex)) AS Unique_Sex_Grouping,\n",
    "      count(distinct(donor_id)) AS Number_Samples,\n",
    "      count(distinct(cell_type)) AS Number_Cell_Types,\n",
    "      count(distinct(treatment)) AS Number_Therapies\n",
    "    FROM\n",
    "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
    "    GROUP BY development_stage\n",
    "    ORDER BY Number_Samples DESC\n",
    "    '''\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ## ScRNAseq example 2\n",
    "    prompt += f\"\"\"\n",
    "    Question: How many unique cell types, sex groupings, cells, and samples\n",
    "    are present in the a specific human stage of development in the MSK\n",
    "    scRNAseq dataset (e.g. 74-year-old)?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "    SELECT\n",
    "      cell_type,\n",
    "      count(distinct(sex)) AS Unique_Sex_Grouping,\n",
    "      count(distinct(iObs)) AS Number_Cells,\n",
    "      count(distinct(donor_id)) AS Number_Samples\n",
    "    FROM\n",
    "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
    "    WHERE\n",
    "      development_stage = '74-year-old human stage'\n",
    "    GROUP BY cell_type\n",
    "    '''\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ## ScRNAseq example 3\n",
    "    prompt += f\"\"\"\n",
    "    Question: How many genes and cells are associated with each sex and\n",
    "    cell type in the MSK scRNAseq dataset for an individual\n",
    "    (e.g. a 74-year-old human stage)?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "    SELECT\n",
    "      sex,\n",
    "      Cell_Type,\n",
    "      count(distinct(feature_name)) AS Number_Genes,\n",
    "      count(distinct(iObs)) AS Number_Cells\n",
    "    FROM\n",
    "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
    "    WHERE development_stage = '74-year-old human stage'\n",
    "    GROUP BY sex, Cell_Type\n",
    "    ORDER BY Cell_Type DESC\n",
    "    '''\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    ## ScRNAseq example 4\n",
    "    prompt += f\"\"\"\n",
    "    Question: How many genes and cells are there in each Seurat Cluster\n",
    "    for males and females of the 'epithelial cell' type in the specific\n",
    "    human stage (here 74-year-old)?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "    SELECT\n",
    "      sex,\n",
    "      clusters,\n",
    "      Cell_Type,\n",
    "      count(distinct(feature_name)) AS Number_Genes,\n",
    "      count(distinct(iObs)) AS Number_Cells\n",
    "    FROM\n",
    "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
    "    WHERE development_stage = '74-year-old human stage' AND Cell_Type = 'epithelial cell'\n",
    "    GROUP BY sex, clusters, Cell_Type\n",
    "    ORDER BY clusters ASC\n",
    "    '''\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "   ## ScRNAseq example 5\n",
    "    prompt += f\"\"\"\n",
    "    Question: How do the average expression values for genes differ between\n",
    "    male and female epithelial cells in a specific cluster, and which genes\n",
    "    show the greatest differences (here cluster 41 of the 74-year-old\n",
    "    human stage)?\n",
    "    Answer:\n",
    "    '''\n",
    "    query = '''\n",
    "    SELECT\n",
    "      A.feature_name,\n",
    "      A.avg_counts_clust10 AS female_avg_counts,\n",
    "      B.avg_counts_clust10 AS male_avg_counts,\n",
    "      A.avg_counts_clust10 - B.avg_counts_clust10 AS mean_diff\n",
    "    FROM (\n",
    "      SELECT\n",
    "        feature_name,\n",
    "        AVG(X_value) AS avg_counts_clust10\n",
    "      FROM\n",
    "        `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
    "      WHERE development_stage = '74-year-old human stage' AND Cell_Type = 'epithelial cell' AND clusters = '41' AND sex = 'female'\n",
    "      GROUP BY feature_name\n",
    "    ) AS A\n",
    "    INNER JOIN (\n",
    "      SELECT\n",
    "        feature_name,\n",
    "        AVG(X_value) AS avg_counts_clust10\n",
    "      FROM\n",
    "        `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
    "      WHERE development_stage = '74-year-old human stage' AND Cell_Type = 'epithelial cell' AND clusters = '41' AND sex = 'male'\n",
    "      GROUP BY feature_name\n",
    "    ) AS B\n",
    "    ON A.feature_name = B.feature_name\n",
    "    ORDER BY mean_diff DESC\n",
    "    '''\n",
    "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
    "    '''\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    # Add the dynamic user question and the placeholder for the answer\n",
    "    prompt += f\"\"\"\n",
    "    Question: {question}\n",
    "    Answer:\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    # Model information and version\n",
    "    vertexai.init(project=\"isb-cgc-external-004\", location=\"us-central1\")\n",
    "    model = GenerativeModel(\n",
    "        \"gemini-1.5-flash-002\",\n",
    "    )\n",
    "\n",
    "    # Responses\n",
    "    responses = model.generate_content(\n",
    "        [prompt],\n",
    "        generation_config=generation_config,\n",
    "        safety_settings=safety_settings,\n",
    "        stream=False,\n",
    "        # tools='code_execution'\n",
    "    )\n",
    "\n",
    "    # for response in responses:\n",
    "    #     print(response.text, end=\"\")\n",
    "\n",
    "    pattern = r'```python\\n(.*?)\\n```'\n",
    "\n",
    "    # Extract the code\n",
    "    match = re.search(pattern, responses.text, re.DOTALL)\n",
    "\n",
    "    if match:\n",
    "        python_code = match.group(1).strip()\n",
    "        print(python_code)\n",
    "        return(python_code)\n",
    "\n",
    "    else:\n",
    "\n",
    "        print(responses)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "crawl4ai",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
